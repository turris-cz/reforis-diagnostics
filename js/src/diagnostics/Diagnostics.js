/*
 * Copyright (C) 2019-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";

import {
    useAPIGet,
    withSpinnerOnSending,
    withErrorMessage,
    formFieldsSize,
} from "foris";

import API_URLs from "API";

import Modules from "./Modules";
import ReportsTable from "./ReportsTable";

export default function Diagnostics() {
    const [getReportsResponse, getReports] = useAPIGet(API_URLs.reports);
    useEffect(() => {
        getReports();
    }, [getReports]);

    return (
        <>
            <h1>{_("Diagnostics")}</h1>
            <p>
                {_(
                    "Generate a diagnostics report to simplify debugging some problems related to the router's functionality."
                )}
            </p>
            <div className={formFieldsSize}>
                <Modules onReload={getReports} />
            </div>
            <div className={formFieldsSize}>
                <h2>{_("Reports")}</h2>
                <ReportsTableWithErrorAndSpinner
                    apiState={getReportsResponse.state}
                    reports={getReportsResponse.data}
                    onReload={getReports}
                />
            </div>
        </>
    );
}

const ReportsTableWithErrorAndSpinner = withSpinnerOnSending(
    withErrorMessage(ReportsTable)
);
