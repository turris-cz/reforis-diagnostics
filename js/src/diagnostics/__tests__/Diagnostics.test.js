/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import {
    render, waitForElement, waitForElementToBeRemoved, fireEvent, getByText, getByTestId
} from "foris/testUtils/customTestRender";
import { mockJSONError } from "foris/testUtils/network";
import mockAxios from 'jest-mock-axios';

import Diagnostics from "../Diagnostics";

describe("<Diagnostics />", () => {
    let componentContainer;

    beforeEach(() => {
        const { container } = render(<Diagnostics />);
        componentContainer = container;
    });

    it("should render modules and reports", async () => {
        expect(getByText(componentContainer, "Diagnostics")).toBeDefined();

        expect(mockAxios.get).toBeCalledWith("/reforis/diagnostics/api/modules", expect.anything());
        // Response to GET modules
        mockAxios.mockResponse({data: {modules: ["foobar_module"]}});

        expect(mockAxios.get).toBeCalledWith("/reforis/diagnostics/api/reports", expect.anything());
        // Response to GET reports
        mockAxios.mockResponse({data: [{diag_id: "1234", path: "/tmp/1234", status: "ready"}]});

        await waitForElement(() => getByText(componentContainer, "Reports"));
        expect(componentContainer).toMatchSnapshot();
    });

    it("should handle error on loading reports", async () => {
        expect(getByText(componentContainer, "Diagnostics")).toBeDefined();

        expect(mockAxios.get).toBeCalledWith("/reforis/diagnostics/api/modules", expect.anything());
        // Response to GET modules
        mockAxios.mockResponse({data: {modules: ["foobar_module"]}});

        expect(mockAxios.get).toBeCalledWith("/reforis/diagnostics/api/reports", expect.anything());
        // Response to GET reports
        mockJSONError();

        await waitForElement(() => getByText(componentContainer, "Reports"));
        expect(componentContainer).toMatchSnapshot();
    });

    it("should reload reports table when new report is generated", async () => {
        // Load initial data
        mockAxios.mockResponse({data: {modules: ["foobar_module"]}});
        mockAxios.mockResponse({data: [{diag_id: "1234", path: "/tmp/1234", status: "ready"}]});
        await waitForElement(() => getByText(componentContainer, "Reports"));

        // Generate new report
        fireEvent.click(getByText(componentContainer, "foobar_module"));
        fireEvent.click(getByText(componentContainer, "Generate report"));

        expect(mockAxios.post).toBeCalledWith(
            "/reforis/diagnostics/api/reports", {modules: ["foobar_module"]}, expect.anything()
        );
        // Response to POST reports
        mockAxios.mockResponse({data: {diag_id: "5678"}});

        // Spinner appears
        await waitForElementToBeRemoved(() => getByTestId(componentContainer, "reports-table"));

        expect(mockAxios.get).toBeCalledWith("/reforis/diagnostics/api/reports", expect.anything());
        // Response to GET reports
        mockAxios.mockResponse({data: [{diag_id: "1234", path: "/tmp/1234", status: "ready"}, {diag_id: "5678", path: "/tmp/5678", status: "ready"}]});

        // New report appears
        await waitForElement(() => getByText(componentContainer, "5678"));
        expect(componentContainer).toMatchSnapshot();
    });
});
