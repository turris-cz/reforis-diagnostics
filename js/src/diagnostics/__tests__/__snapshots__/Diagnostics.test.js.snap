// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Diagnostics /> should handle error on loading reports 1`] = `
<div>
  <h1>
    Diagnostics
  </h1>
  <p>
    Generate a diagnostics report to simplify debugging some problems related to the router's functionality.
  </p>
  <h3>
    Modules
  </h3>
  <p>
    Select modules which you want to include in the report.
  </p>
  <form
    class="col-sm-12 offset-lg-1 col-lg-10"
  >
    <div
      style="font-weight: bold;"
    >
      <div
        class="form-group"
      >
        <div
          class="custom-control custom-checkbox "
        >
          <input
            class="custom-control-input"
            id="1"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="1"
          >
            Select all
          </label>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-sm-12 col-md-6 col-lg-4"
      >
        <div
          class="form-group"
        >
          <div
            class="custom-control custom-checkbox "
          >
            <input
              class="custom-control-input"
              id="2"
              type="checkbox"
            />
            <label
              class="custom-control-label"
              for="2"
            >
              foobar_module
            </label>
          </div>
        </div>
      </div>
    </div>
    <button
      class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
      disabled=""
      type="submit"
    >
       
       
      Generate report
    </button>
  </form>
  <h3
    class="mt-3"
  >
    Reports
  </h3>
  <p
    class="text-center text-danger"
  >
    An error occurred during loading reports
  </p>
</div>
`;

exports[`<Diagnostics /> should reload reports table when new report is generated 1`] = `
<div>
  <h1>
    Diagnostics
  </h1>
  <p>
    Generate a diagnostics report to simplify debugging some problems related to the router's functionality.
  </p>
  <h3>
    Modules
  </h3>
  <p>
    Select modules which you want to include in the report.
  </p>
  <form
    class="col-sm-12 offset-lg-1 col-lg-10"
  >
    <div
      style="font-weight: bold;"
    >
      <div
        class="form-group"
      >
        <div
          class="custom-control custom-checkbox "
        >
          <input
            class="custom-control-input"
            id="1"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="1"
          >
            Select all
          </label>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-sm-12 col-md-6 col-lg-4"
      >
        <div
          class="form-group"
        >
          <div
            class="custom-control custom-checkbox "
          >
            <input
              class="custom-control-input"
              id="2"
              type="checkbox"
            />
            <label
              class="custom-control-label"
              for="2"
            >
              foobar_module
            </label>
          </div>
        </div>
      </div>
    </div>
    <button
      class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
      type="submit"
    >
       
       
      Generate report
    </button>
  </form>
  <h3
    class="mt-3"
  >
    Reports
  </h3>
  <table
    class="table table-hover"
    data-testid="reports-table"
  >
    <tbody>
      <tr>
        <th
          scope="col"
        >
          ID
        </th>
        <th
          aria-label="Download"
          scope="col"
        />
        <th
          aria-label="Delete"
          scope="col"
        />
      </tr>
      <tr>
        <td
          class="align-middle"
        >
          1234
        </td>
        <td
          class="text-center"
        >
          <a
            class="btn btn-primary"
            download=""
            href="/reforis/diagnostics/api/reports/1234/contents"
          >
            Download
          </a>
        </td>
        <td
          class="text-right"
        >
          <button
            class="btn btn-danger"
            type="button"
          >
             
             
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td
          class="align-middle"
        >
          5678
        </td>
        <td
          class="text-center"
        >
          <a
            class="btn btn-primary"
            download=""
            href="/reforis/diagnostics/api/reports/5678/contents"
          >
            Download
          </a>
        </td>
        <td
          class="text-right"
        >
          <button
            class="btn btn-danger"
            type="button"
          >
             
             
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`<Diagnostics /> should render modules and reports 1`] = `
<div>
  <h1>
    Diagnostics
  </h1>
  <p>
    Generate a diagnostics report to simplify debugging some problems related to the router's functionality.
  </p>
  <h3>
    Modules
  </h3>
  <p>
    Select modules which you want to include in the report.
  </p>
  <form
    class="col-sm-12 offset-lg-1 col-lg-10"
  >
    <div
      style="font-weight: bold;"
    >
      <div
        class="form-group"
      >
        <div
          class="custom-control custom-checkbox "
        >
          <input
            class="custom-control-input"
            id="1"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="1"
          >
            Select all
          </label>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-sm-12 col-md-6 col-lg-4"
      >
        <div
          class="form-group"
        >
          <div
            class="custom-control custom-checkbox "
          >
            <input
              class="custom-control-input"
              id="2"
              type="checkbox"
            />
            <label
              class="custom-control-label"
              for="2"
            >
              foobar_module
            </label>
          </div>
        </div>
      </div>
    </div>
    <button
      class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
      disabled=""
      type="submit"
    >
       
       
      Generate report
    </button>
  </form>
  <h3
    class="mt-3"
  >
    Reports
  </h3>
  <table
    class="table table-hover"
    data-testid="reports-table"
  >
    <tbody>
      <tr>
        <th
          scope="col"
        >
          ID
        </th>
        <th
          aria-label="Download"
          scope="col"
        />
        <th
          aria-label="Delete"
          scope="col"
        />
      </tr>
      <tr>
        <td
          class="align-middle"
        >
          1234
        </td>
        <td
          class="text-center"
        >
          <a
            class="btn btn-primary"
            download=""
            href="/reforis/diagnostics/api/reports/1234/contents"
          >
            Download
          </a>
        </td>
        <td
          class="text-right"
        >
          <button
            class="btn btn-danger"
            type="button"
          >
             
             
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
